### PR TITLE
Fixed #124 Add support for apigateway v2 payload

### LINKED ISSUE #124: Support format version 2 for HTTP API's
https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-develop-integrations-lambda.html

Extended the example to deploy an HTTP API and dump its event, got one like this, plus error from trying to parse format version 1:

```
START RequestId: 863fef15-563d-498f-8d32-a4a5737f014f Version: $LATEST                                                                                                                                                                                                                                                                                                                                
{                                                                                                                                                                                                                                                                                                                                                                                                     
  "cookies": [                                                                                                                                                                                                                                                                                                                                                                                        
    "pN=7",                                                                                                                                                                                                                                                                                                                                                                                           
  ],                                                                                                                                                                                                                                                                                                                                                                                                  
  "headers": {                                                                                                                                                                                                                                                                                                                                                                                        
    "accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8",                                                                                                                                                                                                                                                                                                           
    "accept-encoding": "gzip, deflate, br",                                                                                                                                                                                                                                                                                                                                                           
    "accept-language": "en-GB,en;q=0.5",                                                                                                                                                                                                                                                                                                                                                              
    "content-length": "0",                                                                                                                                                                                                                                                                                                                                                                            
    "dnt": "1",                                                                                                                                                                                                                                                                                                                                                                                       
    "host": "s219lqqg4f.execute-api.eu-central-1.amazonaws.com",                                                                                                                                                                                                                                                                                                                                      
    "upgrade-insecure-requests": "1",                                                                                                                                                                                                                                                                                                                                                                 
    "user-agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.14; rv:73.0) Gecko/20100101 Firefox/73.0",                                                                                                                                                                                                                                                                                               
    "x-amzn-trace-id": "Root=1-5e6b67ad-2d4a5560cf6fb7a02780c3a0",                                                                                                                                                                                                                                                                                                                                    
    "x-forwarded-for": "91.109.238.159",                                                                                                                                                                                                                                                                                                                                                              
    "x-forwarded-port": "443",                                                                                                                                                                                                                                                                                                                                                                        
    "x-forwarded-proto": "https"                                                                                                                                                                                                                                                                                                                                                                      
  },                                                                                                                                                                                                                                                                                                                                                                                                  
  "isBase64Encoded": true,                                                                                                                                                                                                                                                                                                                                                                            
  "rawPath": "/",                                                                                                                                                                                                                                                                                                                                                                                     
  "rawQueryString": "",                                                                                                                                                                                                                                                                                                                                                                               
  "requestContext": {                                                                                                                                                                                                                                                                                                                                                                                 
    "accountId": "931198139388",                                                                                                                                                                                                                                                                                                                                                                      
    "apiId": "s219lqqg4f",                                                                                                                                                                                                                                                                                                                                                                            
    "domainName": "s219lqqg4f.execute-api.eu-central-1.amazonaws.com",                                                                                                                                                                                                                                                                                                                                
    "domainPrefix": "s219lqqg4f",                                                                                                                                                                                                                                                                                                                                                                     
    "http": {                                                                                                                                                                                                                                                                                                                                                                                         
      "method": "GET",                                                                                                                                                                                                                                                                                                                                                                                
      "path": "/",                                                                                                                                                                                                                                                                                                                                                                                    
      "protocol": "HTTP/1.1",                                                                                                                                                                                                                                                                                                                                                                         
      "sourceIp": "91.109.238.159",                                                                                                                                                                                                                                                                                                                                                                   
      "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.14; rv:73.0) Gecko/20100101 Firefox/73.0"                                                                                                                                                                                                                                                                                               
    },                                                                                                                                                                                                                                                                                                                                                                                                
    "requestId": "JU0jFiJxliAEJHg=",                                                                                                                                                                                                                                                                                                                                                                  
    "routeId": null,                                                                                                                                                                                                                                                                                                                                                                                  
    "routeKey": "$default",                                                                                                                                                                                                                                                                                                                                                                           
    "stage": "$default",                                                                                                                                                                                                                                                                                                                                                                              
    "time": "13/Mar/2020:10:59:57 +0000",                                                                                                                                                                                                                                                                                                                                                             
    "timeEpoch": 1584097197293                                                                                                                                                                                                                                                                                                                                                                        
  },                                                                                                                                                                                                                                                                                                                                                                                                  
  "routeKey": "$default",                                                                                                                                                                                                                                                                                                                                                                             
  "version": "2.0"                                                                                                                                                                                                                                                                                                                                                                                    
}                                                                                                                                                                                                                                                                                                                                                                                                     
[ERROR] KeyError: 'httpMethod' Traceback (most recent call last):   File "/var/task/testapp/wsgi.py", line 16, in lambda_handler     return apig_wsgi_handler(event, context)   File "/var/task/apig_wsgi.py", line 37, in handler     environ = get_environ(event, context, binary_support=binary_support)   File "/var/task/apig_wsgi.py", line 50, in get_environ     method = event["httpMethod"] 
```

### PR DESCRIPTION
**Pull request by mneil**
Check the event version before building the environment. Use
the unpack operator to copy the original event to the
v1 event object. Map missing keys needed by the environment
builder from the v2 object.

Values like request_context are passed unchanged. The multivalue
headers and query string parameters are mapped back into strings
then split back up into their original formats. It's inefficient
and could be improved but ultimately leaves with unchanged logic
for parsing the an identical end result.

### CODE REVIEW

File: src/apig_wsgi.py [hunk 1]
```diff
@@ -48,6 +49,36 @@ def handler(event, context):
...
+
+
+def get_event_from_version(event):
+    if event.get("version") == "1.0":
+        return event
+
+    headers = event.get("headers", {})
+    multi_value_headers = {}
+    for header, value in headers.items():
+        multi_value_headers[header] = value.split(",")
```
Comment (adamchainz) [inline src/apig_wsgi.py:66]: this is unused

**Review by adamchainz — CHANGES_REQUESTED** [review]
As I understand, your approach rewrites the v2 event to the v1 format. This is rather backwards looking. I'd rather do the opposite.

Also any change needs to update the documentation and changelog.

File: src/apig_wsgi.py (hunk 1 above)
Comment (mneil) [reply src/apig_wsgi.py:66]: Good catch! I missed this. I'll tear it out.

**Comment by mneil** [comment]
🤔 I don't believe I'm rewriting the payload from v2 to v1. I'm taking the v2 payload and coercing the event to have some of the properties from v1 so that the rest of the code in `get_environ` doesn't need to change. If I added the missing parts from v1 to v2 I'd need to additionally modify all the values looked up in `get_environ`.

For example `apig_wsgi.request_context ` contains the event requestContext payload of v2 only. The `apig_wsgi.full_event` contains the entire v2 event with a couple additional properties that were in v1. (all original v2 properties are there unchanged though).

I'll remove the unused code. And you're right that I didn't add documentation for the payload. I will add documentation so that this is better understood. Then we can discuss any impact of the structure and if there are additional changes that should be made?

**Comment by adamchainz** [comment]
@mneil Thanks for starting this. I've taken it over by copying your commits into #174 and carrying on from there.

I split into two implementations, both for environ and response creation. Version 2 is *quite* different and it was getting confusing massaging the v2 event into a "semi v1" state for passing through the original `get_environ` function. It also prevented making full use of all the features, and introduced a bug around multiple cookies being set. Some of the differences:

* The raw query string is available in the `rawQueryString` key, so the logic to reconstitute it is not required.
* Requests contain `cookies` which need combining into a `Cookie` header -the `HTTP_COOKIE` wsgi environ key
* Multiple headers are not supported in responses
* `Set-Cookie` response headers need splitting into the `cookies` response key
* `REMOTE_ADDR`, `REQUEST_METHOD`, and `SERVER_PROTOCOL` can be directly pulled from `requestContext.http` 
* Response binary support is always enabled
* Multiple `set-cookie` headers cannot be sent - multiple cookies should instead be sent through the `cookies` response key.

**Comment by adamchainz** [comment]
Released in version 2.9.0 🥳 Let me know how it works for you!

**Pull request closed without merging**